apps/src/lib/util/speech.js

```diff
@@ -22,38 +20,149 @@ export function textToSpeech(
+            Sounds.getSingleton().onSpeechFinished();
+          }
+        },
+        {
+          language: language,
+          text: text,
+          gender: gender,
+          profanityFoundCallback: profaneWords => {
+            asyncOutputWarning(
+              i18n.textToSpeechProfanity({
+                profanityCount: profaneWords.length,
+                profaneWords: profaneWords.join(', ')
+              })
+            );
+          }
+        }
+      );
+    }
+  } else {
+    let ssml = `<speak version="1.0" xmlns="https://www.w3.org/2001/10/synthesis" xml:lang="en-US"><voice name="${voice}">${text}</voice></speak>`;
+    const url = `https://${region}.tts.speech.microsoft.com/cognitiveservices/v1`;
+
+    var request = new XMLHttpRequest();
+    const resultPromise = new Promise(async function(resolve, reject) {
+      let profaneWords = await findProfanity(text, languageCode);
+      if (profaneWords && profaneWords.length > 0) {
+        if (!Array.isArray(profaneWords)) {
+          profaneWords = [profaneWords];
+        }
+        asyncOutputWarning(
+          i18n.textToSpeechProfanity({
+            profanityCount: profaneWords.length,
+            profaneWords: profaneWords.join(', ')
+          })
+        );
+        Sounds.getSingleton().registerTextBytes(
+          language,
+          text,
+          true,
+          profaneWords
+        );
+        resolve(PROFANITY_FOUND);
+      }
+      request.onreadystatechange = function() {
+        // Waiting for readyState to be 4 meaning that the request is done
+        if (request.readyState === 4) {
+          if (request.status >= 200 && request.status < 300) {
+            resolve(request.response);
+          } else {
+            reject({
+              status: request.status,
+              statusText: request.statusText
+            });
+          }
+        }
+      };
+
+      request.open('POST', url, true);
+      request.responseType = 'arraybuffer';
+      request.setRequestHeader('Authorization', 'Bearer ' + token);
+      request.setRequestHeader('Content-Type', 'application/ssml+xml');
+      request.setRequestHeader(
+        'X-Microsoft-OutputFormat',
+        'audio-16khz-32kbitrate-mono-mp3'
+      );
+      request.send(ssml);
+    });
+    Sounds.getSingleton().addPromiseToSpeechQueue(
+      resultPromise,
+      {
         volume: 1.0,
         loop: false,
         forceHTML5: forceHTML5,
-        allowHTML5Mobile: true
-      });
-
-      synthesizer.close();
-    },
-    error => {
-      console.warn(error);
-      synthesizer.close();
-    }
-  );
+        allowHTML5Mobile: true,
+        onEnded: () => {
+          Sounds.getSingleton().onSpeechFinished();
+        }
+      },
+      {
+        language: language,
+        text: text,
+        hasProfanity: false,
+        profaneWords: undefined,
+        gender: gender
+      }
+    );
+    Sounds.getSingleton().registerTextBytes(
+      language,
+      text,
+      false,
```

Comment:
nit: add a comment for boolean literal parameters like `false /* hasProfanity */` to make it easier to follow